jmespath-core/src/main/java/io/burt/jmespath/function/SortByFunction.java

```diff
@@ -0,0 +1,57 @@
+package io.burt.jmespath.function;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Comparator;
+
+import io.burt.jmespath.Adapter;
+import io.burt.jmespath.node.JmesPathNode;
+
+public class SortByFunction extends JmesPathFunction {
+  public SortByFunction() {
+    super(2, 2);
+  }
+
+  @Override
+  protected <T> T internalCall(final Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    T array = arguments.get(0).value();
+    final JmesPathNode expression = arguments.get(1).expression();
+    if (arguments.get(0).isExpression()) {
+      throw new ArgumentTypeException(name(), "array of objects", "expression");
+    } else if (!adapter.isArray(array)) {
+      throw new ArgumentTypeException(name(), "array of objects", adapter.typeOf(array));
+    }
+    if (arguments.get(1).isValue()) {
+      throw new ArgumentTypeException(name(), "expression", adapter.typeOf(arguments.get(1).value()));
+    }
+    Map<T, T> sortedTransformed = new TreeMap<>(adapter);
+    Iterator<T> elements = adapter.toList(array).iterator();
+    if (elements.hasNext()) {
+      T element = elements.next();
+      T transformedElement = expression.evaluate(adapter, element);
+      boolean expectNumbers = true;
+      if (adapter.isString(transformedElement)) {
+        expectNumbers = false;
+      } else if (!adapter.isNumber(transformedElement)) {
+        throw new ArgumentTypeException(name(), "number or string", adapter.typeOf(transformedElement));
+      }
+      sortedTransformed.put(transformedElement, element);
```

Comment:
Will this not overwrite based on sort-key, effectively keeping only one of the elements that should sort equal?

In principle, I think it would preferable to generate an array-list of `(transformedElement, element)` tuples, and sort them in place, and then remap the elements. In general, this should be faster/better than putting things in a tree.